### PR TITLE
Fixes #2405 by excluding already migrated tasks in subsequent migrati…

### LINKED ISSUE #2405: Migration of ZK State to 0.13/0.14 does not work
Started some tasks in 0.11 and restarted on master. Got following stack trace:

marathon [2015-10-12 16:19:56,348] ERROR Migration failed! (mesosphere.marathon.state.Migration:97)
marathon java.lang.RuntimeException: while migrating storage to 0.13
marathon 	at mesosphere.marathon.state.Migration$$anonfun$migrations$4$$anonfun$apply$2.applyOrElse(Migration.scala:62)
marathon 	at mesosphere.marathon.state.Migration$$anonfun$migrations$4$$anonfun$apply$2.applyOrElse(Migration.scala:61)
marathon 	at scala.runtime.AbstractPartialFunction.apply(AbstractPartialFunction.scala:36)
marathon 	at scala.util.Failure$$anonfun$recover$1.apply(Try.scala:216)
marathon 	at scala.util.Try$.apply(Try.scala:192)
marathon 	at scala.util.Failure.recover(Try.scala:216)
marathon 	at scala.concurrent.Future$$anonfun$recover$1.apply(Future.scala:324)
marathon 	at scala.concurrent.Future$$anonfun$recover$1.apply(Future.scala:324)
marathon 	at scala.concurrent.impl.CallbackRunnable.run(Promise.scala:32)
marathon 	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
marathon 	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
marathon 	at java.lang.Thread.run(Thread.java:745)
marathon Caused by: mesosphere.marathon.StoreCommandFailedException: Could not fetch MarathonTaskState with key: app:app.8f61d32e-70e6-11e5-a216-7a090c3edc79
marathon 	at mesosphere.marathon.state.MarathonStore$$anonfun$mesosphere$marathon$state$MarathonStore$$exceptionTransform$1.applyOrElse(MarathonStore.scala:75)
marathon 	at mesosphere.marathon.state.MarathonStore$$anonfun$mesosphere$marathon$state$MarathonStore$$exceptionTransform$1.applyOrElse(MarathonStore.scala:74)
marathon 	... 10 more
marathon Caused by: com.google.protobuf.InvalidProtocolBufferException: Protocol message end-group tag did not match expected tag.
marathon 	at com.google.protobuf.InvalidProtocolBufferException.invalidEndTag(InvalidProtocolBufferException.java:94)
marathon 	at com.google.protobuf.CodedInputStream.checkLastTagWas(CodedInputStream.java:124)
marathon 	at com.google.protobuf.AbstractParser.parsePartialFrom(AbstractParser.java:143)
marathon 	at com.google.protobuf.AbstractParser.parseFrom(AbstractParser.java:176)
marathon 	at com.google.protobuf.AbstractParser.parseFrom(AbstractParser.java:188)
marathon 	at com.google.protobuf.AbstractParser.parseFrom(AbstractParser.java:193)
marathon 	at com.google.protobuf.AbstractParser.parseFrom(AbstractParser.java:49)
marathon 	at mesosphere.marathon.Protos$MarathonTask.parseFrom(Protos.java:8377)
marathon 	at mesosphere.marathon.state.MarathonTaskState.mergeFromProto(MarathonTaskState.scala:15)
marathon 	at mesosphere.marathon.state.MarathonTaskState.mergeFromProto(MarathonTaskState.scala:9)
marathon 	at mesosphere.marathon.state.MarathonStore.mesosphere$marathon$state$MarathonStore$$stateFromBytes(MarathonStore.scala:79)
marathon 	at mesosphere.marathon.state.MarathonStore$$anonfun$fetch$1$$anonfun$apply$2.apply(MarathonStore.scala:35)
marathon 	at mesosphere.marathon.state.MarathonStore$$anonfun$fetch$1$$anonfun$apply$2.apply(MarathonStore.scala:33)
marathon 	at scala.Option.map(Option.scala:146)
marathon 	at mesosphere.marathon.state.MarathonStore$$anonfun$fetch$1.apply(MarathonStore.scala:33)
marathon 	at mesosphere.marathon.state.MarathonStore$$anonfun$fetch$1.apply(MarathonStore.scala:33)
marathon 	at scala.util.Success$$anonfun$map$1.apply(Try.scala:237)
marathon 	at scala.util.Try$.apply(Try.scala:192)
marathon 	at scala.util.Success.map(Try.scala:237)
marathon 	at scala.concurrent.Future$$anonfun$map$1.apply(Future.scala:235)
marathon 	at scala.concurrent.Future$$anonfun$map$1.apply(Future.scala:235)
marathon 	... 4 more

### PR DESCRIPTION
**Pull request by meichstedt**
…on attempts.

- Migrate tasks sequentially instead of in parallel
- only migrate tasks matching the old key format
- added test to verify that old tasks break migration
- added test to verify already migrated tasks are excluded from migration

Needs to be merged into releasing-0.14 as well.

Note this will only fix the reported subsequent error when trying to migrate already migrated tasks. The underlying problem ist not yet clear, but it might be that the parallel execution of task migrations introduced the problem.

### CODE REVIEW

File: src/main/scala/mesosphere/marathon/state/Migration.scala [hunk 1]
```diff
@@ -322,27 +322,34 @@ class MigrationTo0_13(taskRepository: TaskRepository, store: PersistentStore) {
...
+
+      namesInOldFormat.foldLeft(Future.successful(())) { (f, nextKey) =>
+        f.flatMap(_ => migrateKey(nextKey))
+      }
     }.map { _ =>
       log.info("Completed 0.13 migration")
     }
   }
 
+  // including 0.12, task keys are in format task:app:app.uuid – the taskId is
```
Comment (kolloch) [inline src/main/scala/mesosphere/marathon/state/Migration.scala]: do you mean "task.uuid"?

Comment (kolloch) [inline src/main/scala/mesosphere/marathon/state/Migration.scala]: Nit-picking: An example would be nice. I think the "task:" prefix is literal and "app" and "app.uuid" are place-holders for some actual values. An example would clarify that.

File: src/test/scala/mesosphere/marathon/state/MigrationTo0_13Test.scala [hunk 2]
```diff
@@ -41,6 +42,59 @@ class MigrationTo0_13Test extends MarathonSpec with GivenWhenThen with Matchers
...
+    names should contain (appId.safePath + ":" + task1.getId)
+
+    When("we run the migration")
+    f.migration.migrateTasks().futureValue
+
+    Then("the tasks are stored in paths without duplicated appId")
+    val taskKeys1 = f.taskRepo.tasksKeys(appId).futureValue
+
+    taskKeys1 should not be empty
+    taskKeys1.size should be (1)
```
Comment (kolloch) [inline src/test/scala/mesosphere/marathon/state/MigrationTo0_13Test.scala]: you can combine these statements into "should have size(1)"

File: src/test/scala/mesosphere/marathon/state/MigrationTo0_13Test.scala [hunk 3]
```diff
@@ -41,6 +42,59 @@ class MigrationTo0_13Test extends MarathonSpec with GivenWhenThen with Matchers
...
+    f.legacyTaskStore.store(appId, task2).futureValue
+    f.entityStore.names().futureValue should contain (appId.safePath + ":" + task2.getId)
+
+    And("we run the migration again")
+    f.migration.migrateTasks().futureValue
+
+    Then("Only the second task is considered and the first one does not crash the migration")
+    val taskKeys2 = f.taskRepo.tasksKeys(appId).futureValue
+    taskKeys2 should not be empty
+    taskKeys2.size should be (2)
```
Comment (kolloch) [inline src/test/scala/mesosphere/marathon/state/MigrationTo0_13Test.scala]: see above

**Comment by kolloch** [comment]
Except minor picks looks good to me. Nice tests!

File: src/main/scala/mesosphere/marathon/state/Migration.scala (hunk 1 above)
Comment (meichstedt) [inline src/main/scala/mesosphere/marathon/state/Migration.scala]: Makes sense, I tried to make that more clear and added examples for both old and new formats.

File: src/test/scala/mesosphere/marathon/state/MigrationTo0_13Test.scala (hunk 2 above)
Comment (meichstedt) [inline src/test/scala/mesosphere/marathon/state/MigrationTo0_13Test.scala]: Yay, did that in all places where applicable.

File: src/test/scala/mesosphere/marathon/state/MigrationTo0_13Test.scala (hunk 3 above)
Comment (meichstedt) [inline src/test/scala/mesosphere/marathon/state/MigrationTo0_13Test.scala]: Seen and fixed.

**Comment by meichstedt** [comment]
@kolloch Thanks for your feedback! I incorporated the suggested changes and squashed the commits.

**Comment by bobrik** [comment]
Any chance to get a new RC after merging this? We hit the bug with one of our Marathons and old version likes to have split brains from time to time.

**Comment by aquamatthias** [comment]
Thanks!

**Comment by aquamatthias** [comment]
Hey @bobrik we will release an RC today. Really would appreciate if you could give some feedback for that.

**Pull request merged**